config/dynamic.c

```diff
@@ -16,11 +16,527 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <sys/un.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_cycles.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
 #include "gatekeeper_config.h"
+#include "gatekeeper_launch.h"
+
+/* TODO Get the install-path via Makefile. */
+#define LUA_DY_BASE_DIR     "./lua"
+#define DYNAMIC_CONFIG_FILE "dy.lua"
+
+const char *CLIENT_PROC_ERROR =
+	"function reply() return -1, \"Dynamic Configuration cannot process the request. end\"\n";
+
+const uint16_t MSG_MAX_LEN = (uint16_t)~0U;
+
+/*
+ * Return values:
+ * 	0: Write @nbytes successfully.
+ * 	-1: Connection closed by the client.
+ */
+static int
+write_nbytes(int conn_fd, const char *msg_buff, int nbytes)
+{
+	int send_size;
+	int tot_size = 0;
+
+	while ((send_size = write(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += send_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the client is closed. */
+	if (send_size == 0) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"dyn_cfg: Client disconnected\n");
+		return -1;
+	} else if (send_size < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, uint16_t reply_len)
+{
+	int ret;
+	uint16_t nlen = htons(reply_len);
+
+	/* The first two bytes: the length of the message in network order. */
+	ret = write_nbytes(conn_fd, (char *)&nlen, sizeof(nlen));
+	if (ret != 0)
+		return -1;
+
+	/* Sending the message. */
+	ret = write_nbytes(conn_fd, reply_msg, reply_len);
+	if (ret != 0)
+		return -1;
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+       const char *msg, int msg_len, lua_State *lua_state)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	size_t reply_len;
+	const char *reply_msg;
+
+	/* Load the Lua chunk, and run it. */
+	ret = luaL_loadbuffer(lua_state, msg, msg_len, "message")
```

Comment:
What happens if `msg` is the empty string (i.e. "")? We must gracefully deal with this case.